src/System.Net.HttpListener/src/System/Net/Managed/HttpListenerRequest.Managed.cs

```diff
@@ -643,6 +642,34 @@ public bool IsWebSocketRequest
         {
             get
             {
+                bool foundConnectionUpgradeHeader = false;
+                if (string.IsNullOrEmpty(this.Headers[HttpKnownHeaderNames.Connection]) || string.IsNullOrEmpty(this.Headers[HttpKnownHeaderNames.Upgrade]))
+                {
+                    return false;
+                }
+
+                foreach (string connection in this.Headers.GetValues(HttpKnownHeaderNames.Connection))
+                {
+                    if (string.Compare(connection, HttpKnownHeaderNames.Upgrade, StringComparison.OrdinalIgnoreCase) == 0)
+                    {
+                        foundConnectionUpgradeHeader = true;
+                        break;
+                    }
+                }
+
+                if (!foundConnectionUpgradeHeader)
+                {
+                    return false;
+                }
+
+                foreach (string upgrade in this.Headers.GetValues(HttpKnownHeaderNames.Upgrade))
```

Comment:
Do you know what the implementation of GetValues does?  I'm wondering if it'd be cheaper to loop through all of the headers once, looking for Connection and Upgrade and the relevant string in each, rather than first checking whether they exist, and then separately looping through each.